packages/Meshfree/src/DTK_MultivariatePolynomialBasis.hpp

```diff
@@ -0,0 +1,113 @@
+/****************************************************************************
+ * Copyright (c) 2012-2017 by the DataTransferKit authors                   *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the DataTransferKit library. DataTransferKit is     *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ ****************************************************************************/
+
+#ifndef DTK_MULTIVARIATE_POLYNOMIAL_BASIS_HPP
+#define DTK_MULTIVARIATE_POLYNOMIAL_BASIS_HPP
+
+#include <Kokkos_Array.hpp>
+#include <Kokkos_Macros.hpp>
+
+namespace DataTransferKit
+{
+
+struct Linear
+{
+};
+
+struct Quadratic
+{
+};
+
+namespace Details
+{
+
+template <typename Basis, int DIM>
+struct Traits
+{
+    static KOKKOS_INLINE_FUNCTION int constexpr size();
+};
+
+template <>
+struct Traits<Linear, 3>
+{
+    static KOKKOS_INLINE_FUNCTION int constexpr size() { return 4; }
+};
+
+template <>
+struct Traits<Quadratic, 3>
+{
+    static KOKKOS_INLINE_FUNCTION int constexpr size() { return 10; }
+};
+
+template <>
+struct Traits<Linear, 2>
+{
+    static KOKKOS_INLINE_FUNCTION int constexpr size() { return 3; }
+};
+
+template <>
+struct Traits<Quadratic, 2>
+{
+    static KOKKOS_INLINE_FUNCTION int constexpr size() { return 6; }
+};
+
+} // end namespace Details
+
+template <typename Basis, int DIM>
+struct MultivariatePolynomialBasis
+{
+    static KOKKOS_INLINE_FUNCTION int constexpr size()
+    {
+        return Details::Traits<Basis, DIM>::size();
+    }
+    template <typename Point>
+    KOKKOS_INLINE_FUNCTION Kokkos::Array<double, size()>
+    operator()( Point const &p ) const;
+};
+
+// NOTE: For now relying on Point::operator[]( int i ) to access the coordinates
+// which make it possible to use various types such as DTK::Point or
+// Kokkos::Array<double, DIM>
+
+template <>
+template <typename Point>
+KOKKOS_INLINE_FUNCTION Kokkos::Array<double, 4>
```

Comment:
Shouldn't you replace this `4` with the appropriate `size()`?

Reply:
I would have to replace by `MultivariatePolynomialBasis<Linear, 3>::size()` so don't really want to.

Reply:
Anyway, I did change it.